File: services/documents/dpo_report.py
"""Б2 («самое массивное» из раздела «ИИ чат-бот»): отчёт по ДПО из выгрузки 1С:ЗиК.

Вход: xlsx-выгрузка «ПК за год» (шапка с параметрами, затем таблица: Физическое
лицо, Должность, Подразделение, Иерархия, Категория должности, Учреждение,
Вид образования, Группа, даты, Номер документа, Примечание, Тема, Часов, …).
ВАЖНО: 1С группирует строки по человеку — ФИО заполнено только в ПЕРВОЙ строке
блока, дальше forward-fill.

Выход: word-отчёт по структуре образца «ДПО за 2023»: все числа считаются
ДЕТЕРМИНИРОВАННО из таблицы (в отчёте недопустимы «примерные» цифры LLM).
"""
from __future__ import annotations

import re
from collections import Counter, defaultdict
from datetime import datetime
from pathlib import Path
from typing import Any, Iterable

from sqlalchemy.orm import Session

from config import settings
from data.my_documents import MyDocuments
from data.users import User
from utils.logger import logger

# Запрос «отчёт по ДПО» в чате.
DPO_REQUEST_RE = re.compile(
    r"(отч[её]т\w*[^.\n]{0,50}(\bдпо\b|повышени\w*\s+квалификац))"
    r"|((\bдпо\b|повышени\w*\s+квалификац)[^.\n]{0,50}отч[её]т)",
    re.IGNORECASE,
)

_HEADER_MARKERS = ("физическое лицо", "категория должности")
_SHORT_HOURS = 16  # порог «краткосрочной» программы

# Категории должностей → группы образца отчёта (считаются по УНИКАЛЬНЫМ людям).
_CATEGORY_GROUPS: tuple[tuple[str, tuple[str, ...], str], ...] = (
    ("ППС", ("ППС",), "профессорско-преподавательский состав (ППС)"),
    ("ПС и ПР", ("ПС", "ПР"), "педагогические работники СПО и СОО (ПС и ПР)"),
    ("НР", ("НР", "НТР"), "научные и научно-технические работники (НР, НТР)"),
    ("АУП", ("АУП", "АУПН"), "административно-управленческий персонал (АУП, АУПН)"),
    ("УВП", ("УВП", "ПУП"), "учебно-вспомогательный персонал (УВП, ПУП)"),
    ("ИТР", ("ИТР", "ПРОП"), "инженерно-технические работники (ИТР, ПрОП)"),
    ("АХП", ("АХП", "АХПН"), "административно-хозяйственный персонал (АХП, АХПН)"),
)

# Виды образования (по записям с часами >= 16, как в образце).
_KIND_LABELS = (
    ("Профессиональная переподготовка", "по программам профессиональной переподготовки"),
    ("Профессиональное обучение", "по программам профессионального обучения"),
    ("Повышение квалификации", "по программам повышения квалификации"),
)

# Обязательные программы (перечень образца) → ключи поиска в «Теме».
_MANDATORY = (
    ("Оказание первой помощи", re.compile(r"перв\w*\s+помощ", re.I)),
    ("Противодействие коррупции", re.compile(r"коррупц", re.I)),
    ("Обучение инвалидов и лиц с ОВЗ", re.compile(r"инвалид|ограниченными возможностями|\bОВЗ\b", re.I)),
    ("Охрана труда", re.compile(r"охран\w*\s+труда", re.I)),
    ("Контрактная система в сфере закупок", re.compile(r"закупок|контрактн\w*\s+систем", re.I)),
    ("Гражданская оборона и защита от ЧС", re.compile(r"гражданск\w*\s+оборон", re.I)),
    ("Информационно-образовательная среда", re.compile(r"информационно-образовательн", re.I)),
)

_YEAR_RE = re.compile(r"\b(20\d{2})\b")


def analyze_dpo_rows(rows: Iterable[tuple]) -> dict[str, Any]:
    """Агрегация выгрузки. rows — строки листа (кортежи значений), включая шапку."""
    header_idx: dict[str, int] = {}
    records: list[dict[str, Any]] = []
    current_fio: str | None = None
    param_year: int | None = None

    for row in rows:
        cells = ["" if c is None else str(c).strip() for c in row]
        if not any(cells):
            continue
        low = [c.lower() for c in cells]
        if not header_idx:
            # Параметры до шапки: «Период завершения: 31.12.2023» → год отчёта
            joined = " ".join(cells)
            if "период" in joined.lower():
                m = _YEAR_RE.search(joined)
                if m:
                    param_year = int(m.group(1))
            if any(mk in low for mk in _HEADER_MARKERS):
                header_idx = {name: i for i, name in enumerate(low) if name}
            continue

        def col(name: str) -> str:
            i = header_idx.get(name)
            return cells[i] if i is not None and i < len(cells) else ""

        # forward-fill ФИО (1С показывает его только в первой строке блока)
        fio = col("физическое лицо")
        if fio:
            current_fio = fio
        if not (col("вид образования") or col("тема") or col("категория должности")):
            continue  # служебная строка

        hours_raw = col("часов").replace(",", ".")
        try:
            hours = float(hours_raw) if hours_raw else None
        except ValueError:
            hours = None

        records.append({
            "fio": current_fio or f"(без ФИО #{len(records)})",
            "category": col("категория должности").upper().replace("РОП", "РОП"),
            "kind": col("вид образования"),
            "group": col("группа"),
            "theme": col("тема"),
            "org": col("учереждение") or col("учреждение"),
            "hours": hours,
            "issued": col("дата выдачи"),
        })

    if not records:
        raise ValueError(
            "Не удалось разобрать выгрузку: не найдена шапка таблицы "
            "(«Физическое лицо», «Категория должности», …)"
        )

    people: dict[str, list[dict]] = defaultdict(list)
    for r in records:
        people[r["fio"]].append(r)

    long_recs = [r for r in records if r["hours"] is not None and r["hours"] >= _SHORT_HOURS]
    short_recs = [r for r in records if r["hours"] is not None and r["hours"] < _SHORT_HOURS]

    kind_counts = Counter(r["kind"] for r in long_recs)
    kinds = [(label, kind_counts.get(kind, 0)) for kind, label in _KIND_LABELS]

    # Категории — по уникальным людям (категория человека = самая частая у его записей)
    person_cat: dict[str, str] = {}
    for fio, recs in people.items():
        cats = Counter(r["category"] for r in recs if r["category"])
        person_cat[fio] = cats.most_common(1)[0][0] if cats else ""
    cat_people: list[tuple[str, int]] = []
    for _key, codes, label in _CATEGORY_GROUPS:
        n = sum(1 for c in person_cat.values() if c in codes)
        cat_people.append((label, n))
    known_codes = {c for _k, codes, _l in _CATEGORY_GROUPS for c in codes}
    other_people = sum(1 for c in person_cat.values() if c and c not in known_codes)

    # Формы обучения — по людям (человек может попасть в несколько форм)
    def _people_in(group_re: re.Pattern) -> int:
        return sum(
            1 for recs in people.values()
            if any(group_re.search(r["group"] or "") for r in recs)
        )

    forms = {
        "internal": _people_in(re.compile(r"внутривуз", re.I)),
        "external": _people_in(re.compile(r"иные", re.I)),
        "internship": _people_in(re.compile(r"стажиров", re.I)),
    }

    mandatory = []
    for label, rx in _MANDATORY:
        n_people = sum(1 for recs in people.values() if any(rx.search(r["theme"] or "") for r in recs))
        if n_people:
            mandatory.append((label, n_people))

    theme_people: dict[str, set] = defaultdict(set)
    for r in records:
        if r["theme"]:
            theme_people[r["theme"]].add(r["fio"])
    top_themes = sorted(theme_people.items(), key=lambda kv: len(kv[1]), reverse=True)[:10]

    year = param_year
    if not year:
        years = Counter(
            int(m.group(1)) for r in records
            if (m := _YEAR_RE.search(r["issued"] or ""))
        )
        year = years.most_common(1)[0][0] if years else datetime.now().year

    multi = sum(1 for recs in people.values() if len(recs) >= 2)
    return {
        "year": year,
        "total_records": len(records),
        "total_people": len(people),
        "multi_program_people": multi,
        "total_programs": len(theme_people),
        "long_events": len(long_recs),
        "short_events": len(short_recs),
        "kinds": kinds,                      # [(подпись, число мероприятий ≥16ч)]
        "categories": cat_people,            # [(подпись, человек)]
        "categories_other": other_people,
        "forms": forms,
        "mandatory": mandatory,              # [(программа, человек)]
        "top_themes": [(t, len(p)) for t, p in top_themes],
    }


def analyze_dpo_xlsx(path: str | Path) -> dict[str, Any]:
    from openpyxl import load_workbook

    wb = load_workbook(str(path), read_only=True, data_only=True)
    try:
        ws = wb.worksheets[0]
        return analyze_dpo_rows(ws.iter_rows(values_only=True))
    finally:
        wb.close()


def build_report_text(stats: dict[str, Any]) -> str:
    """Текст отчёта по структуре образца «ДПО за 2023». Все числа — из таблицы."""
    y = stats["year"]
    p: list[str] = []
    p.append(f"Отчет по ДПО за {y} год")
    p.append(
        f"1. Дополнительное профессиональное образование работников университета в {y} году "
        "осуществлялось в соответствии с Порядком организации дополнительного профессионального "
        "образования и внутриорганизационного обучения работников университета, разделом 8 "
        "Коллективного договора ТИУ, разделом 9 Трудового кодекса Российской Федерации и "
        "Федеральным законом от 29.12.2012 № 273-ФЗ «Об образовании в Российской Федерации»."
    )
    p.append(
        f"В {y} году было организовано дополнительное профессиональное образование "
        f"{stats['total_people']} работников по {stats['total_programs']} программам обучения "
        f"(при этом {stats['multi_program_people']} человек — по 2 и более программам)."
    )
    kinds_lines = "\n".join(f"{n} — {label};" for label, n in stats["kinds"])
    p.append(
        f"Всего проведено {stats['long_events']} обучающих мероприятий по программам "
        f"дополнительного профессионального образования (от 16 часов), в том числе:\n{kinds_lines}"
    )
    if stats["short_events"]:
        p.append(f"Проведено {stats['short_events']} краткосрочных программ (до 16 часов).")

    cat_lines = "\n".join(
        f"{label} – {n} человек;" for label, n in stats["categories"] if n
    )
    if stats.get("categories_other"):
        cat_lines += f"\nиные категории – {stats['categories_other']} человек;"
    p.append(f"По категориям должностей обучение проходили:\n{cat_lines}")

    f = stats["forms"]
    p.append(
        "При организации повышения квалификации использовались разные формы обучения:\n"
        f"внутривузовское повышение квалификации (по программам ИДДО) – {f['internal']} человек;\n"
        f"курсы повышения квалификации, программы профессиональной переподготовки в других "
        f"образовательных организациях – {f['external']} человек;\n"
        f"стажировки – {f['internship']} человек."
    )
    if stats["mandatory"]:
        mand = "\n".join(f"{label} – {n} человек;" for label, n in stats["mandatory"])
        p.append(
            f"В {y} году повышение квалификации работников осуществлялось по программам, "
            "входящим в перечень обязательного обучения в соответствии с действующим "
            f"законодательством, в том числе:\n{mand}"
        )
    if stats["top_themes"]:
        top = "\n".join(f"«{t[:120]}» – {n} человек;" for t, n in stats["top_themes"])
        p.append(f"Наиболее массовые программы обучения {y} года:\n{top}")
    p.append(
        "[Разделы о программах по приоритетным направлениям развития университета, "
        "бесплатных онлайн-программах и выполнении плана ВОО заполняются вручную — "
        "этих данных нет в выгрузке 1С.]"
    )
    return "\n\n".join(p)


def render_dpo_docx(stats: dict[str, Any], body_text: str) -> Path:
    from docx import Document
    from docx.enum.text import WD_ALIGN_PARAGRAPH
    from docx.shared import Pt

    doc = Document()
    style = doc.styles["Normal"]
    style.font.name = "Times New Roman"
    style.font.size = Pt(12)

    paras = re.split(r"\n\s*\n+", body_text)
    h = doc.add_paragraph()
    h.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = h.add_run(paras[0])
    run.bold = True
    run.font.size = Pt(14)
    for para in paras[1:]:
        lines = para.split("\n")
        p0 = doc.add_paragraph(lines[0])
        p0.alignment = WD_ALIGN_PARAGRAPH.JUSTIFY
        for extra in lines[1:]:
            li = doc.add_paragraph(extra)
            li.paragraph_format.left_indent = Pt(24)

    settings.docs_generated.mkdir(parents=True, exist_ok=True)
    out = settings.docs_generated / f"dpo_report_{stats['year']}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.docx"
    doc.save(str(out))
    return out


def create_dpo_report(db: Session, user: User, xlsx_path: str | Path) -> tuple[MyDocuments, str, dict]:
    """Полный цикл: xlsx → агрегаты → текст → docx → «Мои документы»."""
    stats = analyze_dpo_xlsx(xlsx_path)
    text = build_report_text(stats)
    path = render_dpo_docx(stats, text)
    rec = MyDocuments(
        user_id=user.id,
        title=f"Отчёт по ДПО за {stats['year']} год",
        template_key="dpo_report",
        file_path=str(path),
        progress=100,
        status="ready",
        fields={"year": stats["year"], "total_people": stats["total_people"],
                "long_events": stats["long_events"]},
    )
    db.add(rec)
    db.commit()
    db.refresh(rec)
    logger.info("[DPO] отчёт создан: doc={} ({} записей, {} человек)",
                rec.id, stats["total_records"], stats["total_people"])
    return rec, text, stats

File: app/lib/docs/dpo-report.ts
import 'server-only';
import { Prisma } from '@prisma/client';
import { prisma } from '@/lib/db';
import { buildDocx, type DocxPara } from './docx';
import { DocValueError, mostCommon, ru, saveGenerated, timestamp } from './common';
import { toDocsPath } from '@/lib/news';

/**
 * Б2: отчёт по ДПО из выгрузки 1С:ЗиК «ПК за период».
 * Порт services/documents/dpo_report.py.
 *
 * ВАЖНО: 1С группирует строки по человеку — ФИО заполнено только в ПЕРВОЙ
 * строке блока, дальше forward-fill. Все числа считаются ДЕТЕРМИНИРОВАННО из
 * таблицы: в отчёте недопустимы «примерные» цифры LLM.
 */

/** Триггер чат-команды «отчёт по ДПО» / «отчёт о повышении квалификации». */
export const DPO_REQUEST_RE = ru(
  '(отч[её]т\\w*[^.\\n]{0,50}(\\bдпо\\b|повышени\\w*\\s+квалификац))' +
    '|((\\bдпо\\b|повышени\\w*\\s+квалификац)[^.\\n]{0,50}отч[её]т)',
  'i'
);

const HEADER_MARKERS = ['физическое лицо', 'категория должности'];
const SHORT_HOURS = 16; // порог «краткосрочной» программы

// Категории должностей → группы образца отчёта (считаются по УНИКАЛЬНЫМ людям).
const CATEGORY_GROUPS: [string, string[], string][] = [
  ['ППС', ['ППС'], 'профессорско-преподавательский состав (ППС)'],
  ['ПС и ПР', ['ПС', 'ПР'], 'педагогические работники СПО и СОО (ПС и ПР)'],
  ['НР', ['НР', 'НТР'], 'научные и научно-технические работники (НР, НТР)'],
  ['АУП', ['АУП', 'АУПН'], 'административно-управленческий персонал (АУП, АУПН)'],
  ['УВП', ['УВП', 'ПУП'], 'учебно-вспомогательный персонал (УВП, ПУП)'],
  ['ИТР', ['ИТР', 'ПРОП'], 'инженерно-технические работники (ИТР, ПрОП)'],
  ['АХП', ['АХП', 'АХПН'], 'административно-хозяйственный персонал (АХП, АХПН)'],
];

// Виды образования (по записям с часами >= 16, как в образце).
const KIND_LABELS: [string, string][] = [
  ['Профессиональная переподготовка', 'по программам профессиональной переподготовки'],
  ['Профессиональное обучение', 'по программам профессионального обучения'],
  ['Повышение квалификации', 'по программам повышения квалификации'],
];

// Обязательные программы (перечень образца) → ключи поиска в «Теме».
const MANDATORY: [string, RegExp][] = [
  ['Оказание первой помощи', ru('перв\\w*\\s+помощ', 'i')],
  ['Противодействие коррупции', /коррупц/i],
  ['Обучение инвалидов и лиц с ОВЗ', ru('инвалид|ограниченными возможностями|\\bОВЗ\\b', 'i')],
  ['Охрана труда', ru('охран\\w*\\s+труда', 'i')],
  ['Контрактная система в сфере закупок', ru('закупок|контрактн\\w*\\s+систем', 'i')],
  ['Гражданская оборона и защита от ЧС', ru('гражданск\\w*\\s+оборон', 'i')],
  ['Информационно-образовательная среда', /информационно-образовательн/i],
];

const YEAR_RE = ru('\\b(20\\d{2})\\b');

/** float() из Python: «16,5» уже приведена к точке, мусор даёт null. */
function pyFloat(s: string): number | null {
  const t = s.trim();
  if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(t)) return null;
  const v = Number(t);
  return Number.isFinite(v) ? v : null;
}

interface DpoRecord {
  fio: string;
  category: string;
  kind: string;
  group: string;
  theme: string;
  org: string;
  hours: number | null;
  issued: string;
}

export interface DpoStats {
  year: number;
  total_records: number;
  total_people: number;
  multi_program_people: number;
  total_programs: number;
  long_events: number;
  short_events: number;
  kinds: [string, number][];
  categories: [string, number][];
  categories_other: number;
  forms: { internal: number; external: number; internship: number };
  mandatory: [string, number][];
  top_themes: [string, number][];
}

/** Агрегация выгрузки. rows — строки листа, включая шапку. */
export function analyzeDpoRows(rows: string[][]): DpoStats {
  let headerIdx: Map<string, number> | null = null;
  const records: DpoRecord[] = [];
  let currentFio: string | null = null;
  let paramYear: number | null = null;

  for (const row of rows) {
    const cells = row.map((c) => (c ?? '').trim());
    if (!cells.some((c) => c)) continue;
    const low = cells.map((c) => c.toLowerCase());

    if (!headerIdx) {
      // Параметры до шапки: «Период завершения: 31.12.2023» → год отчёта
      const joined = cells.join(' ');
      if (joined.toLowerCase().includes('период')) {
        const m = YEAR_RE.exec(joined);
        if (m) paramYear = Number.parseInt(m[1], 10);
      }
      if (HEADER_MARKERS.some((mk) => low.includes(mk))) {
        headerIdx = new Map();
        low.forEach((name, i) => {
          if (name) headerIdx!.set(name, i);
        });
      }
      continue;
    }

    const col = (name: string): string => {
      const i = headerIdx!.get(name);
      return i !== undefined && i < cells.length ? cells[i] : '';
    };

    // forward-fill ФИО (1С показывает его только в первой строке блока)
    const fio = col('физическое лицо');
    if (fio) currentFio = fio;
    if (!(col('вид образования') || col('тема') || col('категория должности'))) continue;

    const hoursRaw = col('часов').replace(/,/g, '.');
    records.push({
      fio: currentFio || `(без ФИО #${records.length})`,
      category: col('категория должности').toUpperCase(),
      kind: col('вид образования'),
      group: col('группа'),
      theme: col('тема'),
      org: col('учереждение') || col('учреждение'),
      hours: hoursRaw ? pyFloat(hoursRaw) : null,
      issued: col('дата выдачи'),
    });
  }

  if (!records.length) {
    throw new DocValueError(
      'Не удалось разобрать выгрузку: не найдена шапка таблицы ' +
        '(«Физическое лицо», «Категория должности», …)'
    );
  }

  const people = new Map<string, DpoRecord[]>();
  for (const r of records) {
    const list = people.get(r.fio);
    if (list) list.push(r);
    else people.set(r.fio, [r]);
  }

  const longRecs = records.filter((r) => r.hours !== null && r.hours >= SHORT_HOURS);
  const shortRecs = records.filter((r) => r.hours !== null && r.hours < SHORT_HOURS);

  const kindCounts = new Map<string, number>();
  for (const r of longRecs) kindCounts.set(r.kind, (kindCounts.get(r.kind) ?? 0) + 1);
  const kinds: [string, number][] = KIND_LABELS.map(([kind, label]) => [label, kindCounts.get(kind) ?? 0]);

  // Категории — по уникальным людям (категория человека = самая частая у его записей)
  const personCat = new Map<string, string>();
  for (const [fio, recs] of people) {
    personCat.set(fio, mostCommon(recs.map((r) => r.category).filter((c) => c)) ?? '');
  }
  const categories: [string, number][] = CATEGORY_GROUPS.map(([, codes, label]) => [
    label,
    [...personCat.values()].filter((c) => codes.includes(c)).length,
  ]);
  const knownCodes = new Set(CATEGORY_GROUPS.flatMap(([, codes]) => codes));
  const categoriesOther = [...personCat.values()].filter((c) => c && !knownCodes.has(c)).length;

  // Формы обучения — по людям (человек может попасть в несколько форм)
  const peopleIn = (rx: RegExp): number =>
    [...people.values()].filter((recs) => recs.some((r) => rx.test(r.group || ''))).length;
  const forms = {
    internal: peopleIn(/внутривуз/i),
    external: peopleIn(/иные/i),
    internship: peopleIn(/стажиров/i),
  };

  const mandatory: [string, number][] = [];
  for (const [label, rx] of MANDATORY) {
    const n = [...people.values()].filter((recs) => recs.some((r) => rx.test(r.theme || ''))).length;
    if (n) mandatory.push([label, n]);
  }

  const themePeople = new Map<string, Set<string>>();
  for (const r of records) {
    if (!r.theme) continue;
    const set = themePeople.get(r.theme);
    if (set) set.add(r.fio);
    else themePeople.set(r.theme, new Set([r.fio]));
  }
  const topThemes = [...themePeople.entries()]
    .sort((a, b) => b[1].size - a[1].size)
    .slice(0, 10);

  let year = paramYear;
  if (!year) {
    const years: number[] = [];
    for (const r of records) {
      const m = YEAR_RE.exec(r.issued || '');
      if (m) years.push(Number.parseInt(m[1], 10));
    }
    year = mostCommon(years) ?? new Date().getFullYear();
  }

  const multi = [...people.values()].filter((recs) => recs.length >= 2).length;
  return {
    year,
    total_records: records.length,
    total_people: people.size,
    multi_program_people: multi,
    total_programs: themePeople.size,
    long_events: longRecs.length,
    short_events: shortRecs.length,
    kinds,
    categories,
    categories_other: categoriesOther,
    forms,
    mandatory,
    top_themes: topThemes.map(([t, p]) => [t, p.size] as [string, number]),
  };
}

/** Текст отчёта по структуре образца «ДПО за 2023». Все числа — из таблицы. */
export function buildReportText(stats: DpoStats): string {
  const y = stats.year;
  const p: string[] = [];
  p.push(`Отчет по ДПО за ${y} год`);
  p.push(
    `1. Дополнительное профессиональное образование работников университета в ${y} году ` +
      'осуществлялось в соответствии с Порядком организации дополнительного профессионального ' +
      'образования и внутриорганизационного обучения работников университета, разделом 8 ' +
      'Коллективного договора ТИУ, разделом 9 Трудового кодекса Российской Федерации и ' +
      'Федеральным законом от 29.12.2012 № 273-ФЗ «Об образовании в Российской Федерации».'
  );
  p.push(
    `В ${y} году было организовано дополнительное профессиональное образование ` +
      `${stats.total_people} работников по ${stats.total_programs} программам обучения ` +
      `(при этом ${stats.multi_program_people} человек — по 2 и более программам).`
  );
  const kindsLines = stats.kinds.map(([label, n]) => `${n} — ${label};`).join('\n');
  p.push(
    `Всего проведено ${stats.long_events} обучающих мероприятий по программам ` +
      `дополнительного профессионального образования (от 16 часов), в том числе:\n${kindsLines}`
  );
  if (stats.short_events) {
    p.push(`Проведено ${stats.short_events} краткосрочных программ (до 16 часов).`);
  }

  let catLines = stats.categories
    .filter(([, n]) => n)
    .map(([label, n]) => `${label} – ${n} человек;`)
    .join('\n');
  if (stats.categories_other) {
    catLines += `\nиные категории – ${stats.categories_other} человек;`;
  }
  p.push(`По категориям должностей обучение проходили:\n${catLines}`);

  const f = stats.forms;
  p.push(
    'При организации повышения квалификации использовались разные формы обучения:\n' +
      `внутривузовское повышение квалификации (по программам ИДДО) – ${f.internal} человек;\n` +
      'курсы повышения квалификации, программы профессиональной переподготовки в других ' +
      `образовательных организациях – ${f.external} человек;\n` +
      `стажировки – ${f.internship} человек.`
  );
  if (stats.mandatory.length) {
    const mand = stats.mandatory.map(([label, n]) => `${label} – ${n} человек;`).join('\n');
    p.push(
      `В ${y} году повышение квалификации работников осуществлялось по программам, ` +
        'входящим в перечень обязательного обучения в соответствии с действующим ' +
        `законодательством, в том числе:\n${mand}`
    );
  }
  if (stats.top_themes.length) {
    const top = stats.top_themes.map(([t, n]) => `«${t.slice(0, 120)}» – ${n} человек;`).join('\n');
    p.push(`Наиболее массовые программы обучения ${y} года:\n${top}`);
  }
  p.push(
    '[Разделы о программах по приоритетным направлениям развития университета, ' +
      'бесплатных онлайн-программах и выполнении плана ВОО заполняются вручную — ' +
      'этих данных нет в выгрузке 1С.]'
  );
  return p.join('\n\n');
}

async function renderDpoDocx(stats: DpoStats, bodyText: string): Promise<string> {
  const blocks = bodyText.split(/\n\s*\n+/);
  const paras: DocxPara[] = [
    { runs: [{ text: blocks[0], bold: true, sizePt: 14 }], align: 'center' },
  ];
  for (const block of blocks.slice(1)) {
    const lines = block.split('\n');
    paras.push({ runs: [{ text: lines[0] }], align: 'justify' });
    for (const extra of lines.slice(1)) {
      paras.push({ runs: extra ? [{ text: extra }] : [], leftIndentPt: 24 });
    }
  }
  return saveGenerated(`dpo_report_${stats.year}_${timestamp()}.docx`, buildDocx(paras));
}

/** Полный цикл: агрегаты → текст → docx → «Мои документы». */
export async function createDpoReport(userId: number, rows: string[][]) {
  const stats = analyzeDpoRows(rows);
  const text = buildReportText(stats);
  const filePath = await renderDpoDocx(stats, text);
  const rec = await prisma.my_documents.create({
    data: {
      user_id: userId,
      title: `Отчёт по ДПО за ${stats.year} год`,
      template_key: 'dpo_report',
      file_path: toDocsPath(filePath),
      progress: 100,
      status: 'ready',
      fields: {
        year: stats.year,
        total_people: stats.total_people,
        long_events: stats.long_events,
      } as Prisma.InputJsonValue,
      is_pii: true, // списки обученных работников — документ не хранится
    },
  });
  return { rec, text, stats };
}
